### DISCUSSION TITLE
Steps for a clean complete reinstall of Docker and Reach on native Ubuntu + bypass of docker-compose package

### DISCUSSION THREAD

**Discussion by julzaa** [reach-sh/reach-lang#1300] [General]
Hey everyone!

My story is that I initially had a problem with my installation of Reach (and Docker) on my Ubuntu Jammy OS (native, not WSL), because when I entered
`reach compile overview/index.rsh` I got:
> reach: /app/tmp/cli/2022-07-22T16-27-15Z-Zjmb/out.sh: openFile: permission denied (Permission denied)

But on my Windows 10 system (with Docker already installed) it worked fine, so something was clearly off.

After frankly too many hours of tinkering and trying to make it work, and having this other error one time (copy-pasted from someone else):
> reach: /app/tmp/cli/2022-07-02T20-54-15Z-H4ig/out.sh: openFile: does not exist (No such file or directory)

(I even tried using SSH but it turns out VS Code doesn't support SSH + WSL), I decided to make a fresh reinstall of Docker and Reach, so removing all containers, images, data, etc. of Docker (**careful with that!**), and wrote a protocol to follow for a system like mine, if anyone is blocked by some issue the (kind of hardcore) solution may very well be it.
Again, _BE CAREFUL_, it will wipe everything associated with Docker and Reach on a x64 system running Ubuntu Jammy natively, please understand at least the broad concept of every command in this file if you decide to go this path to make it work, I will not take any responsibility if anything bad happens to your operating system.

Furthermore in my research I found that the Reach documentation repeated the steps of the Docker installation, so it might cause issues if run again (while having a different type of install, for example), and also found that we don't have to install the docker-compose package, instead we can use the official "switch" that in my understanding maps docker-compose to the new "docker compose" package without having to install Docker Desktop. Reach is using docker-compose (the V1) which is coded in Python, but the new "docker compose" (the V2) is coded in Golang which is substantially faster. The steps are at the end.

I'm not a Linux expert so some optional steps may be redundant or useless.

Anyway, here are these steps I wrote for myself (but that could also be useful to you):

### Manually uninstall the Reach IDE extension on VS Code (by yourself), then remove Reach from the system using these (_CAREFUL_ it will delete your Reach directory created at the beginning and everything in it, including your programs and such):
```
rm -r ~/reach
rm -r ~/.config/reach
```

#### (optional) Manually verify that everything is removed, be wary that there are many false positives
```
sudo find / -name '*reach*'
```

#### (optional) Kill all docker processes
```
systemctl --user stop docker-desktop
top | grep -i docker
killall com.docker.back
pgrep -l 'docker'
sudo pkill docker
ps -aux | grep -i docker
```
	
### If the docker-compose switch was previously installed
```
update-alternatives --remove docker-compose /usr/local/bin/compose-switch
sudo rm /usr/local/bin/compose-switch
```

### List all docker packages to check which ones to remove (it won't display containerd.io)
```
dpkg -l | grep -i docker
```
### Remove all docker-related packages (sources: [1](https://fedingo.com/how-to-uninstall-docker-in-ubuntu/), [2](https://lokarithm.com/2020/05/31/how-to-completely-remove-docker-from-debian-ubuntu-or-your-raspberry-pi/))
```
sudo apt-get purge docker-ce docker-ce-cli docker-ce-rootless-extras docker-compose-plugin docker-desktop docker-scan-plugin containerd.io
sudo apt-get autoremove --purge docker-ce docker-ce-cli docker-ce-rootless-extras docker-compose-plugin docker-desktop docker-scan-plugin containerd.io
sudo apt-get autoclean
```
#### Verify that all packages are gone
```
dpkg -l | grep -i docker
```
#### In case you installed the docker-compose CLI plugin (source: [3](https://docs.docker.com/compose/install/uninstall/#uninstalling-the-docker-compose-cli-plugin))
```
sudo rm -r $DOCKER_CONFIG/cli-plugins/docker-compose /usr/local/lib/docker/cli-plugins/docker-compose
```
#### Remove all other directories and files related to docker (sources: [4](https://docs.docker.com/engine/install/ubuntu/#uninstall-docker-engine), [5](https://docs.docker.com/desktop/install/ubuntu/#prerequisites))
```
sudo rm /usr/local/bin/com.docker.cli ~/.config/systemd/user/docker-desktop.service /etc/apparmor.d/docker
sudo rm -rf /var/lib/docker /etc/docker /var/lib/containerd /var/run/docker.sock /usr/local/bin/docker-compose /etc/docker ~/.docker
sudo groupdel docker
```
```
rm -r ~/.password-store/docker-credential-helpers
sudo rm -r /root/.docker
```

#### Remove keyrings and signed repos (because why not, I'm doing the reverse route of installation)
```
sudo rm /etc/apt/keyrings/docker.gpg /usr/share/keyrings/docker-archive-keyring.gpg /etc/apt/sources.list.d/docker.list
```
#### Verify manually that everything is uninstalled, the grep result of the first command is expected, and terminate the last one after a few seconds using CTRL+C
```
ps -aux | grep -i docker
sudo find / -name '*docker*'
pgrep -l 'docker'
top | grep -i docker
```
	
### Docker Engine (source: [6](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository)), run the last 5 lines together
```
sudo apt-get update
sudo apt-get install \
 ca-certificates \
 curl \
 gnupg \
 lsb-release
```

```
sudo mkdir -p /etc/apt/keyrings
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
```

#### Run these 3 lines together:
```
echo \
 "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
 $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
```

```
sudo apt-get update
sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose-plugin
```

#### This command should not work at this step because "The docker group is created but no users are added to it.", if that's the case run the post-installation steps below and do NOT sudo that:
```
docker run hello-world
```

### Post installation steps (source: [7](https://docs.docker.com/engine/install/linux-postinstall/))
```
sudo groupadd docker
sudo usermod -aG docker $USER
newgrp docker
docker run hello-world
```
	
### Verify that everything is properly installed
```
docker compose version
docker --version
docker version
```
	
### Installation steps for Reach (source: [8](https://docs.reach.sh/quickstart/#qs-linux-install))
```
sudo apt install make
mkdir -p ~/reach && cd ~/reach
curl https://docs.reach.sh/reach -o reach ; chmod +x reach
```

### Command that displays "Reach relies on an installation of docker-compose; do you want to install it?", enter NO
```
./reach version
```

### B - docker-compose switch (compatibility resolver) (sources: [8](https://github.com/docker/compose-switch) found in [9](https://github.com/docker/compose#about-update-and-backward-compatibility))
```
sudo curl -fL https://github.com/docker/compose-switch/releases/latest/download/docker-compose-linux-amd64 -o /usr/local/bin/compose-switch
sudo chmod +x /usr/local/bin/compose-switch
sudo update-alternatives --install /usr/local/bin/docker-compose docker-compose /usr/local/bin/compose-switch 99
update-alternatives --display docker-compose
```

Let me know if this helped!

EDIT: Added _sudo_ for line _update-alternatives --install /usr/local/bin/docker-compose docker-compose /usr/local/bin/compose-switch 99_

**Comment by jeapostrophe** [comment]
Thanks! I believe that the source of your problem the first time was missing this sentence --- https://docs.reach.sh/quickstart/#p_39 --- and those post-install steps --- https://docs.docker.com/engine/install/linux-postinstall/

**Reply by julzaa** [reply]
Thanks for the reply but no I didn't miss this step, I could run `docker run hello-world` or even `docker run -it ubuntu bash` without any issues.
The weird part was that I tried running `docker run hello-world` on my first Docker installation just before doing the post-installation steps, and it worked. It might have been because I installed Docker Desktop before that, maybe another local package on top of an online repo, and messed up the installation order so it could have messed up the permissions (among other things).
It is likely that just executing `sudo groupdel docker` before redoing the post-installation steps, then `rm -r ~/.config/reach` before trying to run reach again would have fixed the issue.

I already reinstalled everything once but without going to such lengths and it didn't work, that's why I made this protocol to be so sure it'd work that I'd be hopeless if it hadn't!
